File: packages/sdk/java/core/src/main/java/dev/bugwatch/Breadcrumb.java
package dev.bugwatch;

import java.time.Instant;
import java.util.Map;

/**
 * Represents a breadcrumb trail entry for debugging context.
 */
public record Breadcrumb(
        String timestamp,
        String type,
        String category,
        String message,
        Map<String, Object> data,
        String level
) {

    /**
     * Creates a new breadcrumb with the current timestamp.
     */
    public static Breadcrumb create(String type, String category, String message, Level level) {
        return new Breadcrumb(
                Instant.now().toString(),
                type,
                category,
                message,
                null,
                level != null ? level.getValue() : null
        );
    }

    /**
     * Creates a new breadcrumb with the current timestamp and additional data.
     */
    public static Breadcrumb create(String type, String category, String message, Map<String, Object> data, Level level) {
        return new Breadcrumb(
                Instant.now().toString(),
                type,
                category,
                message,
                data,
                level != null ? level.getValue() : null
        );
    }
}

File: packages/sdk/php/src/Types/Breadcrumb.php
<?php

declare(strict_types=1);

namespace Bugwatch\Types;

final class Breadcrumb
{
    /**
     * @param array<string, mixed> $data
     */
    public function __construct(
        public readonly float $timestamp,
        public readonly string $type = 'default',
        public readonly string $category = '',
        public readonly string $message = '',
        public readonly array $data = [],
        public readonly Level $level = Level::Info,
    ) {}

    /**
     * @param array<string, mixed> $data
     */
    public static function create(
        string $category,
        string $message = '',
        Level $level = Level::Info,
        string $type = 'default',
        array $data = [],
    ): self {
        return new self(
            timestamp: microtime(true),
            type: $type,
            category: $category,
            message: $message,
            data: $data,
            level: $level,
        );
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return array_filter([
            'timestamp' => $this->timestamp,
            'type' => $this->type,
            'category' => $this->category,
            'message' => $this->message,
            'data' => $this->data ?: null,
            'level' => $this->level->value,
        ], fn (mixed $value): bool => $value !== null);
    }
}
